crates/rlp/src/decode.rs

```diff
@@ -36,360 +30,168 @@ impl<'a> Rlp<'a> {
-        Ok(match u8::decode(buf)? {
-            0 => false,
-            1 => true,
-            _ => return Err(DecodeError::Custom("invalid bool value, must be 0 or 1")),
-        })
-    }
-}
-
-impl<const N: usize> Decodable for [u8; N] {
+impl Decodable for Bytes {
     #[inline]
-    fn decode(from: &mut &[u8]) -> Result<Self, DecodeError> {
-        let h = Header::decode(from)?;
-        if h.list {
-            return Err(DecodeError::UnexpectedList)
-        }
-        if h.payload_length != N {
-            return Err(DecodeError::UnexpectedLength)
-        }
-        if from.remaining() < N {
-            return Err(DecodeError::InputTooShort)
-        }
-
-        let mut to = [0_u8; N];
-        to.copy_from_slice(&from[..N]);
-        from.advance(N);
-
-        Ok(to)
+    fn decode(buf: &mut &[u8]) -> Result<Self> {
+        Header::decode_bytes(buf, false).map(|x| Self::from(x.to_vec()))
     }
 }
 
 impl Decodable for BytesMut {
     #[inline]
-    fn decode(from: &mut &[u8]) -> Result<Self, DecodeError> {
-        let h = Header::decode(from)?;
-        if h.list {
-            return Err(DecodeError::UnexpectedList)
-        }
-        if from.remaining() < h.payload_length {
-            return Err(DecodeError::InputTooShort)
-        }
-        let mut to = BytesMut::with_capacity(h.payload_length);
-        to.extend_from_slice(&from[..h.payload_length]);
-        from.advance(h.payload_length);
-
-        Ok(to)
+    fn decode(buf: &mut &[u8]) -> Result<Self> {
+        Header::decode_bytes(buf, false).map(Self::from)
     }
 }
 
-impl Decodable for Bytes {
+impl Decodable for alloc::string::String {
     #[inline]
-    fn decode(buf: &mut &[u8]) -> Result<Self, DecodeError> {
-        BytesMut::decode(buf).map(BytesMut::freeze)
+    fn decode(buf: &mut &[u8]) -> Result<Self> {
+        Header::decode_str(buf).map(Into::into)
     }
 }
 
-impl Decodable for alloc::string::String {
+impl<T: Decodable> Decodable for alloc::vec::Vec<T> {
     #[inline]
-    fn decode(from: &mut &[u8]) -> Result<Self, DecodeError> {
-        let h = Header::decode(from)?;
-        if h.list {
-            return Err(DecodeError::UnexpectedList)
-        }
-        if from.remaining() < h.payload_length {
-            return Err(DecodeError::InputTooShort)
+    fn decode(buf: &mut &[u8]) -> Result<Self> {
+        let mut bytes = Header::decode_bytes(buf, true)?;
+        let mut vec = Self::new();
+        let payload_view = &mut bytes;
+        while !payload_view.is_empty() {
+            vec.push(T::decode(payload_view)?);
         }
-        let mut to = alloc::vec::Vec::with_capacity(h.payload_length);
-        to.extend_from_slice(&from[..h.payload_length]);
-        from.advance(h.payload_length);
-
-        Self::from_utf8(to).map_err(|_| DecodeError::Custom("invalid string"))
+        Ok(vec)
     }
 }
 
-impl<E: Decodable> Decodable for alloc::vec::Vec<E> {
+#[cfg(feature = "smol_str")]
```

Comment:
does smol_str have a motivating usecase downstream? It's not really a common library

Reply:
From reth-rlp, I'm assuming it's used somewhere there but I haven't verified that